### ISSUE TITLE
Questions regarding the implementation

### ISSUE THREAD

**Issue by yuvaraj91** [hyunwoongko/transformer#3]
1. In this file https://github.com/hyunwoongko/transformer/blob/master/models/model/transformer.py, you define the functions `make_pad_mask` and `make_no_peak_mask`, but it is actually used during training?
2. In this file, https://github.com/hyunwoongko/transformer/blob/master/models/layers/position_wise_feed_forward.py why does your `PositionwiseFeedForward` have extra layers?

**Comment by Yuvaraj91** [comment]
Thank you for your reply :) 

For (2), yes you are right actually. I made a mistake when comparing the `class PositionwiseFeedForward`with these two implentations; in https://github.com/bentrevett/pytorch-seq2seq/blob/master/6%20-%20Attention%20is%20All%20You%20Need.ipynb and http://nlp.seas.harvard.edu/2018/04/03/attention.html.
But now I see that yours is the same, but you just coded it in a different format. 

I have another question, how could we visualise the attention heatmap at the decoder heads, similar to https://github.com/bentrevett/pytorch-seq2seq/blob/master/6%20-%20Attention%20is%20All%20You%20Need.ipynb?

**Comment by Qing-zhan** [comment]
Hi, could you share all the requirements of this repo, like pytorch version etc. Thanks.

**Comment by linghushaoxia** [comment]
> Hi, could you share all the requirements of this repo, like pytorch version etc. Thanks.
those requirements is from pip freeze
`#python 3.8
blis==0.7.11
catalogue==2.0.10
certifi==2024.7.4
charset-normalizer==3.3.2
click==8.1.7
colorama==0.4.6
contourpy==1.1.1
cycler==0.12.1
cymem==2.0.8
de-core-news-sm @ https://github.com/explosion/spacy-models/releases/download/de_core_news_sm-3.3.0/de_core_news_sm-3.3.0-py3-none-any.whl#sha256=505b40810815883aba2687e845b63ea5537c78f7c6dc770ec648f894538c3f4f
en-core-web-sm @ https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.3.0/en_core_web_sm-3.3.0-py3-none-any.whl#sha256=84d7d8059bfbf53c09b39139782f76cd6ac7064851e7799dcc685c06ebf5fd4f
filelock==3.15.4
fonttools==4.53.1
fsspec==2024.6.1
idna==3.7
importlib_resources==6.4.0
Jinja2==3.1.4
kiwisolver==1.4.5
langcodes==3.4.0
language_data==1.2.0
marisa-trie==1.2.0
MarkupSafe==2.1.5
matplotlib==3.7.0
mpmath==1.3.0
murmurhash==1.0.10
networkx==3.1
numpy==1.24.4
packaging==24.1
pathlib_abc==0.1.1
pathy==0.11.0
pillow==10.4.0
preshed==3.0.9
pydantic==1.7.4
pyparsing==3.1.2
python-dateutil==2.9.0.post0
requests==2.32.3
six==1.16.0
smart-open==6.4.0
spacy==3.3.1
spacy-legacy==3.0.12
spacy-loggers==1.0.5
srsly==2.4.8
sympy==1.13.1
thinc==8.0.17
#pip install torch==1.8.1+cpu torchvision==0.9.1+cpu torchaudio==0.8.1 -f https://download.pytorch.org/whl/torch_stable.html
torch==1.8.1+cpu
torchaudio==0.8.1
torchtext==0.9.1
torchvision==0.9.1+cpu
tqdm==4.66.5
typer==0.4.2
typing_extensions==4.12.2
urllib3==2.2.2
wasabi==0.10.1
zipp==3.20.0
`
